### ISSUE TITLE
vibe.web.rest.Collection and vibe.web.auth

### ISSUE THREAD

**Issue by grogancolin** [vibe-d/vibe.d#1922]
👋 

Sorry for the large chunk of code I'm gonna paste here, but it's the smallest I could've gotten it while keeping useful info.

Basically, I have an API that is behind a `@requiresAuth`. and inside that API is a sub API that I'd like to also require authentication, but also use `vibe.web.rest.Collection` on.

However, the sub API fails to receive the `auth` parameter from the `authenticate` call.
Also, as I have a log message in the `authenticate` mixin, I can see that the Sub APIs `authenticate` method is being called at the expected time.

These are the commands I used to test:
```bash
# check if I can make an authenticated request
$ curl -b cookies -XGET http://localhost:8080/test\?num=3 
403 - Forbidden

Not authorized to perform this request

# Login
$ curl -c cookies -XPOST -H "Content-Type: application/json" -d '{"user":"admin","pass":"supersecret"}' localhost:8080/login
true

# check if I can make an authenticated request
$ curl -b cookies -XGET http://localhost:8080/test\?num=3
"admin3"

$ curl -b cookies -XGET http://localhost:8080/items/name\?item=this
400 - Bad Request

Missing non-optional query parameter 'auth'.
```

And now for the code. It's basically an amalgamation of:
http://vibed.org/api/vibe.web.rest/Collection
http://vibed.org/api/vibe.web.auth/
```D
import vibe.vibe;
import vibe.web.auth;
import vibe.web.rest;

void main()
{
    auto settings = new HTTPServerSettings;
    settings.port = 8080;
    //settings.sessionStore = new RedisSessionStore("127.0.0.1", 0);
    settings.sessionStore = new MemorySessionStore();

    auto router = new URLRouter;
    router.registerRestInterface(new API());

    listenHTTP(settings, router);
    runApplication();
}

struct AuthInfo{
    string name;
}

/// The Item API that's using vibe.web.rest.Collection
@requiresAuth
interface I_ItemAPI {
    struct CollectionIndices {
        string item;
    }

    @anyAuth
    string getName(string item, AuthInfo auth);

    mixin CreateAuthFunc;
}

class ItemAPI : I_ItemAPI {
    /+
    // I'd expect 'auth' to be populated with the session, but no such luck
       $ curl -b cookies -XGET http://localhost:8080/items/name\?item=this
       400 - Bad Request

        Missing non-optional query parameter 'auth'.
    +/
    string getName(string item, AuthInfo auth){
        return auth.name ~ item;
    }
}

/// Main API entry point
@requiresAuth
interface I_API {

    /+
        // Logging in with the following and saving the cookies to a file
        $ curl -c cookies -XPOST \
                -H "Content-Type: application/json" \
                -d '{"user":"admin","pass":"supersecret"}' \
                localhost:8080/login
        true
    +/
    @noAuth
    @before!getReq("req")
    @before!getRes("res")
    bool postLogin(string user, string pass, HTTPServerRequest req, HTTPServerResponse res);

    /+
    // This is a test just to ensure the Auth stuff is working. (It does)
       $ curl -b cookies -XGET http://localhost:8080/test\?num=3
        "admin3"
    +/
    @anyAuth
    string getTest(AuthInfo auth, int num);

    // /items/
    Collection!I_ItemAPI items();

    mixin CreateAuthFunc;
}


// can only login with 'admin' / 'supersecret'
// This is stored in a SessionStore
class API : I_API{
    bool postLogin(string user, string pass, HTTPServerRequest req, HTTPServerResponse res){
        if(req.session && req.session.isKeySet("auth")){
            return true;
        }
        if(user == "admin" && pass == "supersecret"){
            auto session = res.startSession();
            session.set("auth", AuthInfo(user));
            return true;
        }
        return false;
    }

    // GET /test?num=5
    // returns: admin5
    string getTest(AuthInfo auth, int num){
        return auth.name ~ num.to!string;
    }

    private ItemAPI m_items;
    this(){
        m_items = new ItemAPI();
    }

    // Send all /items/ calls to sub I_ItemAPI interface
    Collection!I_ItemAPI items() {
        return Collection!I_ItemAPI(m_items);
    }
}

auto getReq(HTTPServerRequest req, HTTPServerResponse _){
    return req;
}

auto getRes(HTTPServerRequest _, HTTPServerResponse res){
    return res;
}

// helper template so I don't have to copy/paste code to all interfaces
mixin template CreateAuthFunc(){
    final AuthInfo authenticate(HTTPServerRequest req, HTTPServerResponse res){
        logInfo("Authenticating from %s", __FUNCTION__);
        if(req.session && req.session.isKeySet("auth")) {
            return req.session.get!AuthInfo("auth");
        }
        else {
            throw new HTTPStatusException(HTTPStatus.forbidden, "Not authorized to perform this request");
        }
    }
}
```

**Comment by grogancolin** [comment]
Could it be that the UDAs aren't being passed on to the item and/or non-item methods?

I don't know the code particularly well, but it looks like it's generating a function to forward the call to the interfaces function here:
https://github.com/vibe-d/vibe.d/blob/master/web/vibe/web/rest.d#L870
https://github.com/vibe-d/vibe.d/blob/master/web/vibe/web/rest.d#L915

And this generated function would need to take the `@auth` attribute as it's what's getting compiled into the rest interface in the end?
Total wild guess here...

**Comment by s-ludwig** [comment]
I didn't realize it, but #1939 actually fixes this, too.

**Comment by grogancolin** [comment]
Ah nice! 
I had numerous attempts at understanding where this is going wrong - I knew it was in some of the compile time magic, just couldn't pinpoint where.

Thanks @s-ludwig !

**Comment by s-ludwig** [comment]
Fixed by #1939.
